src/ArrayInterface.jl

```diff
@@ -35,6 +35,8 @@ Otherwise, return `nothing`.
 known_length(x) = known_length(typeof(x))
 known_length(::Type{<:NTuple{N,<:Any}}) where {N} = N
 known_length(::Type{<:NamedTuple{L}}) where {L} = length(L)
+known_length(::Type{T}) where {T<:Base.Slice} = known_length(parent_type(T))
```

Comment:
> I was thinking a bit about `Base.Slice` in the context of `HybridArrays` and my general conclusion was to treat is the same way as `Colon`. 

We are treating them as traversing the entire axis, so a `Base.Slice` preserves density.

> IIRC there was a case where a statically-known axis ended up being a slice with `Base.OneTo` instead of `SOneTo`

`SOneTo` is preferable, so I'd recommend getting that case to return a `Base.Slice(SOneTo(N))` instead.

> Are there any benefits of using this method instead?

Using which method instead of which alternative?

Currently, `known_length` takes a type and returns `nothing` (if length unknown, e.g. for `Base.Array`s), or returns the known length.
A problem is that we can't really use this within `@generated` functions, lest we face world age issues / defeat the entire purpose of providing an interface others can extend (I just checked and realize I have to fix `_sdsize`).

I'm considering a breaking change where `known_length` will return the `Static` numbers introduced in this PR, which will make it easier to "wrap" `@generated` functions (i.e., the wrapper can safely call `known_length` and friends, and with `Static`, it can pass the information on to the `@generated` function).
Although this works:
```julia
L = known_length(T)
if isnothing(L)
   # handle
else
  Static(L) # do something with this
end
```

Reply:
> `SOneTo` is preferable, so I'd recommend getting that case to return a `Base.Slice(SOneTo(N))` instead.

OK, I'll keep that in mind.

> Using which method instead of which alternative?

In HybridArrays I take into account both array type and types of indices to figure out whether a certain access pattern is statically-sized or not. In some places for example `Colon` has to be special-cased, so the same special-casing can be used for `Slice`, like I did here for example: https://github.com/JuliaArrays/StaticArrays.jl/pull/783/files#diff-c1a2ec8ab9a030d018ed5124a63d38a4R193 . I don't see in what circumstances `known_length(::Type{<:Base.Slice})` would be used that don't treat `Colon` in a special way as well.

> I'm considering a breaking change where `known_length` will return the `Static` numbers introduced in this PR, which will make it easier to "wrap" `@generated` functions (i.e., the wrapper can safely call `known_length` and friends, and with `Static`, it can pass the information on to the `@generated` function).

I support this change. Generated functions that need static size information should get that information as an argument. That's what `StaticArray` does in many places. The downside main I think is potentially increasing compilation times.

Reply:
Just a small reminder that you haven't fixed `_sdsize` yet: https://github.com/SciML/ArrayInterface.jl/pull/61/files#diff-b8b78add44fe2319a1282ed26bc78610R284

Reply:
Oops, that slipped my mind. Added.